### ISSUE TITLE
ECKey: unexpected private key (BigInteger) obtained from a WiF private key

### ISSUE THREAD

**Issue by natzei** [bitcoinj/bitcoinj#1372]
# Problem
The problem arises creating a private key with a BigInteger that exceeds 32 bytes. The WiF representation truncates the private key with unexpected behaviour

The problem is due to the WiF representation (Wallet Import Format) which is 
 ```
  [1 byte version]
  [32 bytes of the private key, i.e. BigInteger bytes]
  [1 byte compression] (optional)
  [4 byte checksum]
```

Due to this representation, when the private key (BigInteger) is greater than 32 bytes it shouldn't be possible:
1. to create a key with a BigInteger
2. or to generate a WiF representation

## Solution
1. _ECKey.fromPrivate(BigInteger b)_ should fail if b exceeds 32 bytes
2. _Utils.bigIntegerToBytes(BigInteger b, int numBytes)_ should fail if b exceeds 32 bytes

# Example
```java
NetworkParameters params = NetworkParameters.fromID(NetworkParameters.ID_MAINNET);

BigInteger good = new BigInteger("5c6dcc4abbbd6ff562346234623464326bf31e37ecbc68cebc82896ae7619fbc", 16);   // exactly 32-bytes
BigInteger bad  = new BigInteger("5c6dcc4abbbd6ff562346234623464326bf31e37ecbc68cebc82896ae7619fbc88", 16); // exactly 33-bytes
BigInteger bad2 = new BigInteger("5c6dcc4abbbd6ff562346234623464326bf31e37ecbc68cebc82896ae7619fbc88888888", 16); // > 33-bytes

ECKey kGood = ECKey.fromPrivate(good);
ECKey kBad =  ECKey.fromPrivate(bad);
ECKey kBad2 = ECKey.fromPrivate(bad2);

assert kGood.getPrivKey().equals(good);
assert kBad.getPrivKey().equals(bad);
assert kBad2.getPrivKey().equals(bad2);

String kGoodWiF = kGood.getPrivateKeyAsWiF(params);     // KzKP2XkH93ytD9SdP7yKBwZERMHhwWTT5TXreWAwBhC1zMjQcBAE
String kBadWiF = kBad.getPrivateKeyAsWiF(params);       // Kzu9JFTWLNChGScJqdzRCxg8SYopsiKkFwDw9GXLVm8LzajSoyhn
String kBad2WiF = kBad2.getPrivateKeyAsWiF(params);     // KzKP2XkH93ytD9SdP7yKBwZERMHhwWTT5TXreWAwBhC1zMjQcBAE

ECKey k5GoodCopy = DumpedPrivateKey.fromBase58(params, kGoodWiF).getKey();
ECKey k5BadCopy =  DumpedPrivateKey.fromBase58(params, kBadWiF).getKey();
ECKey k5Bad2Copy =  DumpedPrivateKey.fromBase58(params, kBad2WiF).getKey();

assert k5GoodCopy.getPrivKey().equals(good);    // 
assert k5BadCopy.getPrivKey().equals(bad);      // k5BadCopy.getPrivKey()  == 6dcc4abbbd6ff562346234623464326bf31e37ecbc68cebc82896ae7619fbc88
assert k5Bad2Copy.getPrivKey().equals(bad2);    // k5Bad2Copy.getPrivKey() == 5c6dcc4abbbd6ff562346234623464326bf31e37ecbc68cebc82896ae7619fbc
```

**Comment by schildbach** [comment]
Thanks! I propose solution 1 via https://github.com/bitcoinj/bitcoinj/pull/1373.

Does anyone think we'll need more than 32 byte ECKeys in Bitcoin soon-ish?

**Comment by Christewart** [comment]
Isn't there something more subtle going on here? `BigInteger`'s are signed therefore you *can* represent a valid 32 byte private key with 33 bytes?

**Comment by schildbach** [comment]
Good point, but isn't BigInteger.bitLength() taking care of that?

**Comment by natzei** [comment]
I think the function `Utils.bigIntegerToBytes` should fail too.
I'm not into the details about newly generated keys (with `new ECKey()`). I don't know if the library you are using might generate such a big number.

**Comment by Christewart** [comment]
@schildbach Yep. That is a handy little function -- I didn't realize it excludes the sign bit. For anyone else reading: https://www.tutorialspoint.com/java/math/biginteger_bitlength.htm

**Comment by schildbach** [comment]
@natzei I agree we should change Utils.bigIntegerToBytes such that it fails if a longer array would be needed. Currently the bytes are simply truncated, which I imagine could be dangerous if the method is used wrong.

**Comment by natzei** [comment]
I think https://github.com/bitcoinj/bitcoinj/pull/1373 closes this issue. Or am I missing something?

**Comment by natzei** [comment]
Solved with commits 38902195 and 3cb1f6c6.
